Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -1883,10 +1883,38 @@ extension BrowserViewController: TabManagerDelegate {
         tab.appStateDelegate = self
     }
 
+    func addTabToRecentTabs(tab: Tab) {
+        if var recentTabs = profile.prefs.dictionaryForKey("recentlyClosedTabs") {
+            var urls = recentTabs["urls"] as! [String]
+            urls.insert(tab.displayURL!.absoluteString, atIndex: 0)
+
+            var titles = recentTabs["titles"] as! [String]
+            titles.insert(tab.displayTitle, atIndex: 0)
+
+            var faviconUrls = recentTabs["faviconUrls"] as! [String]
+            faviconUrls.insert((tab.displayFavicon?.url ?? ""), atIndex: 0)
+
+            recentTabs["urls"] = urls
+            recentTabs["titles"] = titles
+            recentTabs["faviconUrls"] = faviconUrls
+            profile.prefs.setObject(recentTabs, forKey: "recentlyClosedTabs")
+        } else {
+            var recentTabsDict = [String:[String]]()
+            recentTabsDict["urls"] = [tab.displayURL!.absoluteString]
+            recentTabsDict["faviconUrls"] = [tab.displayFavicon?.url ?? ""]
+            recentTabsDict["titles"] = [tab.displayTitle]
+            profile.prefs.setObject(recentTabsDict, forKey: "recentlyClosedTabs")
+        }
+    }
+
     func tabManager(tabManager: TabManager, didRemoveTab tab: Tab) {
         updateTabCountUsingTabManager(tabManager)
         // tabDelegate is a weak ref (and the tab's webView may not be destroyed yet)
         // so we don't expcitly unset it.
+
+        if !AboutUtils.isAboutURL(tab.url) && !tab.isPrivate {
+            addTabToRecentTabs(tab)
```

Comment:
Is the feature supposed to be recently closed tabs or recent tabs? The pref key is named `recentlyClosedTabs` but we're adding the tab when we add or remove one.